jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -15,65 +15,113 @@
  */
 public class RetainableByteBuffer implements Retainable
 {
-    private final ByteBufferPool pool;
     private final ByteBuffer buffer;
-    private final AtomicInteger references;
+    private final AtomicInteger references = new AtomicInteger();
+    private final Consumer<ByteBuffer> releaser;
+    private final AtomicLong lastUpdate = new AtomicLong(System.nanoTime());
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size)
+    RetainableByteBuffer(ByteBuffer buffer, Consumer<ByteBuffer> releaser)
     {
-        this(pool, size, false);
+        this.releaser = releaser;
+        this.buffer = buffer;
     }
 
-    public RetainableByteBuffer(ByteBufferPool pool, int size, boolean direct)
+    public int capacity()
     {
-        this.pool = pool;
-        this.buffer = pool.acquire(size, direct);
-        this.references = new AtomicInteger(1);
+        return buffer.capacity();
     }
 
     public ByteBuffer getBuffer()
     {
         return buffer;
     }
 
-    public int getReferences()
+    public long getLastUpdate()
     {
-        return references.get();
+        return lastUpdate.getOpaque();
     }
 
+    /**
+     * Checks if {@link #retain()} has been called at least one more time than {@link #release()}.
+     * @return true if this buffer is retained, false otherwise.
+     */
+    public boolean isRetained()
+    {
+        return references.get() > 1;
+    }
+
+    public boolean isDirect()
+    {
+        return buffer.isDirect();
+    }
+
+    /**
+     * Increments the retained counter of this buffer. It must be done internally by
+     * the pool right after creation and after each un-pooling.
+     * The reason why this method exists on top of {@link #retain()} is to be able to
+     * have some safety checks that must know why the ref counter is being incremented.
+     */
+    void acquire()
+    {
+        if (references.getAndIncrement() != 0)
+        {
+            references.decrementAndGet();
+            throw new IllegalStateException("re-pooled while still used " + this);
+        }
+    }
+
+    /**
+     * Increments the retained counter of this buffer.
+     */
     @Override
     public void retain()
     {
-        while (true)
+        if (references.getAndIncrement() == 0)
         {
-            int r = references.get();
-            if (r == 0)
-                throw new IllegalStateException("released " + this);
-            if (references.compareAndSet(r, r + 1))
-                break;
+            references.decrementAndGet();
+            throw new IllegalStateException("released " + this);
         }
     }
 
-    public int release()
+    /**
+     * Decrements the retained counter of this buffer.
+     * @return true if the buffer was re-pooled, false otherwise.
+     */
+    public boolean release()
     {
         int ref = references.decrementAndGet();
-        if (ref == 0)
-            pool.release(buffer);
-        else if (ref < 0)
+        if (ref < 0)
```

Comment:
ditto - use an update to avoid changing value when you should not.

Reply:
Good point, I'll do this change.